airbyte-cdk/python/airbyte_cdk/sources/declarative/manifest_declarative_source.py

```diff
@@ -46,37 +50,62 @@ class ManifestDeclarativeSource(DeclarativeSource):
 
     VALID_TOP_LEVEL_FIELDS = {"check", "definitions", "schemas", "spec", "streams", "type", "version"}
 
-    def __init__(self, source_config: ConnectionDefinition, debug: bool = False):
+    def __init__(self, source_config: ConnectionDefinition, debug: bool = False, construct_using_pydantic_models: bool = False):
         """
         :param source_config(Mapping[str, Any]): The manifest of low-code components that describe the source connector
         :param debug(bool): True if debug mode is enabled
         """
         self.logger = logging.getLogger(f"airbyte.{self.name}")
 
+        # Controls whether we build components using the manual handwritten schema and Pydantic models or the legacy flow
+        self.construct_using_pydantic_models = construct_using_pydantic_models
+
+        # For ease of use we don't require the type to be specified at the top level manifest, but it should be included during processing
+        manifest = dict(source_config)
+        if "type" not in manifest:
+            manifest["type"] = "DeclarativeSource"
+
         evaluated_manifest = {}
-        resolved_source_config = ManifestReferenceResolver().preprocess_manifest(source_config, evaluated_manifest, "")
-        self._source_config = resolved_source_config
+        resolved_source_config = ManifestReferenceResolver().preprocess_manifest(manifest, evaluated_manifest, "")
+        propagated_source_config = ManifestComponentTransformer().propagate_types_and_options("", resolved_source_config, {})
+        self._new_source_config = propagated_source_config
+        self._legacy_source_config = resolved_source_config
         self._debug = debug
-        self._factory = DeclarativeComponentFactory()
+        self._legacy_factory = DeclarativeComponentFactory()  # Legacy factory used to instantiate declarative components from the manifest
+        self._constructor = ModelToComponentFactory()  # New factory which converts the manifest to Pydantic models to construct components
 
         self._validate_source()
 
         # Stopgap to protect the top-level namespace until it's validated through the schema
-        unknown_fields = [key for key in self._source_config.keys() if key not in self.VALID_TOP_LEVEL_FIELDS]
+        unknown_fields = [key for key in self._legacy_source_config.keys() if key not in self.VALID_TOP_LEVEL_FIELDS]
```

Comment:
is this check needed or would it be enough to validate the `ConnectorManifest` with additionalProperties = false? would be nice to only define the top level fields once

Reply:
Yep I will add it to the declarative handwritten schema. But while we are keeping the old validation flow, we should leave this block. But during cleanup we will remove this. Will update this w/ a comment